include/rmm/device_uvector.hpp

```diff
@@ -119,6 +119,63 @@ class device_uvector {
   {
   }
 
+  /**
+   * @brief Returns pointer to the specified element
+   *
+   * @throws rmm::out_of_range exception if `element_index >= size()`
+   *
+   * @param element_index Index of the specified element.
+   * @return T* Pointer to the desired element
+   */
+  T* get_pointer_to(std::size_t element_index)
+  {
+    RMM_EXPECTS(
+      element_index <= size(), rmm::out_of_range, "Attempt to access out of bounds element.");
+    return data() + element_index;
+  }
+
+  /**
+   * @brief Copies `v` to the specified element in device memory.
+   *
+   * @note: This function incurs a host to device memcpy and should be used sparingly.
+   *
+   * @throws rmm::out_of_range exception if `element_index >= size()`
+   *
+   * @param element_index Index of the target element
+   * @param v The value to copy to the specified element
+   * @param s The stream on which to perform the copy
+   */
+  void set_element(std::size_t element_index, T const& v, cudaStream_t s)
+  {
+    RMM_CUDA_TRY(
+      cudaMemcpyAsync(get_pointer_to(element_index), &v, sizeof(v), cudaMemcpyDefault, s));
+
+    // TODO: Should this function synchronize? If it doesn't, the danger is that this function can
```

Comment:
I want to make sure we think long and hard about this point.

For instance, it's critical that `set_element` passes `v` by reference (as opposed to by value). Otherwise, `set_element` could return before the copy is complete, and `v` is no longer safe to copy from.

Reply:
Options I see:
1) `set_element` syncs. This ensures correct copy behavior, but imposes a sync (by definition).
2) `set_element` syncs, and we provide a `set_element_async`. Same behavior, but there's an option for async for when the user knows exactly what they want, or needs to set multiple elements followed by a manual sync (or last call is `set_element` sync.
3) pass by value. You're saying this isn't an option, but I'm not sure why. If passed by value, doesn't the function get it's own copy? Pointers being the exception...

async/sync is a hazard of the trade. IMO we let users decide which to use and when, by adding both `set_element` and `set_element_async`. Afaict, This is the approach cuda takes in most cases.

Reply:
re: 3.

Consider 
```
void to_device(int * device_pointer, int v, cudaStream_t s){
   cudaMemcpyAsync(device_pointer, &v, sizeof(v), cudaMemcpyHostToDevice, s);
}
```

This function makes its own local copy of `v` on the function stack. When the function returns, it's stack is popped (including `v`). If the copy hasn't completed before `to_device` returns, then `v` is no longer valid to copy from because it has been destroyed.

Reply:
Indubitably. I was not taking that in to consideration. I would expect cuda to provide the ability to handle the lifetime of an object needing copy. In other languages or frameworks, this can be done by providing a callback with it's own scope, which would live so long as the context keeps that callback alive (aka, until the next sync).

Perhaps `cudaLaunchHostFunc` could facilitate this? The idea being that a call to `set_element_async` would actually enqueue a function to set the element. I have not thought this through in terms of state management or performance.

Reply:
Just realized `cudaLaunchHostFunc` does not allow cuda api calls to be performed from the callback host function.

Reply:
Another ridiculous idea, an api that calls a function and then syncs afterward:

```
template<typename T>
void execute_and_sync(cudaStream_t stream, T func)
{
    func();
    cudaStreamSynchronize(stream);
}
```
```
execute_and_sync(stream, [&](){
    uvec.set_element_async(...);
    uvec.set_element_async(...);
});
```

I think having both `set_element` and `set_element_async` is a good approach, as it gives users options.